support/reftest/reftest

```diff
@@ -0,0 +1,296 @@
+    )
+
+
+def parse_args():
+    root_parser = argparse.ArgumentParser(description=__doc__)
+    subparsers = root_parser.add_subparsers(
+        dest="command", help="reference test data operations"
+    )
+    parser_prepare = subparsers.add_parser(
+        "prepare",
+        help="""
+        prepare reference test data for integration test
+         e.g. $./reftest prepare --bucket dichen-test --table dichen-test
+        """,
+    )
+    parser_prepare.add_argument(
+        "--release-date",
+        default=None,
+        help="Date on which the content will be made available.",
+    )
+    parse_aws_info(parser_prepare)
+    parse_cdn_certs(parser_prepare)
+
+    parser_clean = subparsers.add_parser(
+        "clean",
+        help="""
+        clean reference test data in s3 and dynamodb
+        e.g. $./reftest clean --bucket dichen-test --table dichen-test
+        """,
+    )
+    parse_aws_info(parser_clean)
+
+    return root_parser.parse_args()
+
+
+class RefTestConfig:
+    def __init__(self, prod_cdn_url, test_data):
+        self.prod_cdn_url = prod_cdn_url
+        self.test_data = test_data
+
+
+def load_config():
+    with open("data.yml") as f:
+        config = yaml.load(f, yaml.SafeLoader)
+    return RefTestConfig(config["prod-cdn-url"], config["test_data"],)
+
+
+def download_to_local(url, key_path, cert_path, cacert_path):
+    with requests.get(
+        url,
+        cert=(cert_path, key_path),
+        verify=cacert_path,
+        stream=True,
+        timeout=(30, 30),
+    ) as r:
+        r.raise_for_status()
+        total_size = int(r.headers.get("content-length", 0))
+        t = tqdm(desc=url, total=total_size, unit="iB", unit_scale=True)
+
+        # the file is deleted as soon as it is closed.
+        temp_file = tempfile.NamedTemporaryFile(delete=True)
+        for chunk in r.iter_content(chunk_size=8192):
+            if chunk:  # filter out keep-alive new chunks
+                t.update(len(chunk))
+                temp_file.write(chunk)
+        t.close()
+        return temp_file
+
+
+def prepare(db, s3, config, opt):
+    for item in config.test_data:
+        url = config.prod_cdn_url + item["path"]
+        # download test data to local with a name "NamedTemporaryFile"
+        temp_file = download_to_local(url, opt.key, opt.cert, opt.cacert)
+        # push test data to s3 and dynamodb
+        s3.upload_from_localfile(
+            temp_file.name, item["sha256"], item["content-type"]
+        )
+        db.put_item(item["path"], item["sha256"], opt.release_date)
+        # delete the NamedTemporaryFile
+        temp_file.close()
+
+
+def clean(db, s3, config):
+    for item in config.test_data:
+        key = db.search_object_key_from_web_uri(item["path"])
+        if not key:
+            print("{} not found in table {}".format(item["path"], db.table))
+            continue
+        db.delete_item_from_web_uri(item["path"])
+        s3.delete_from_key(item["sha256"])
```

Comment:
Again, it looks like we'll always delete the "right" file from the s3 bucket but will delete a pseudo-random item from the db table. **_If_** the checksum given to delete_from_key() is accurate, we should be able to also give it to delete_item_from_web_uri() so that we know we're deleting the right table item.

Reply:
@nathanegillett you are talking about a `delete_item_from_checksum()` or `delete_item_from_object_key`, right?

Reply:
Yeah. Deleting items by object_key is 100% accurate, while deleting by web_uri could be a gamble.

Reply:
See my other comments as well. Hopefully it all makes sense together.

Reply:
I think we should delete DynamoDB items using *both* web uri and checksum to find the item for deletion.

I'm not sure we should be deleting stuff from S3 at all.  There is a many-to-one relationship between web URIs and S3 objects, a given object could be referenced by loads of URIs including some not mentioned in reftest.yml.  Particularly for simple file formats like 'listing' files.

Example: https://cdn.redhat.com/content/dist/rhel/server/5/5.7/listing content is

```
i386
ia64
x86_64
```

If you deleted S3 object for this, you would be deleting the object for *every* listing file with that content, of which there will be many, not only the one defined in the testdata.  Similarly if you deleted DynamoDB entries while looking only at the object key and not the URI.

I'm still not really clear on what the desired behavior of 'clean' is here, but if you make it delete S3 objects then it's only safe if the environment is more or less dedicated to hold only this testdata and nothing else.  I didn't expect it to be used that way.  I think it'd be better not to delete the objects, so this can be used on an environment which is also used for other stuff.